File: Python/7py 1/bits_battle.py
"""
|--------------------------------------------------------------------------
| Bits Battle
|--------------------------------------------------------------------------
|
"""

def bits_battle(numbers):
   odd = len(list(filter(lambda i: i == '1', ''.join(list(map(lambda i: "{0:b}".format(i), (filter(lambda i: i%2!=0, numbers))))))))
   even = len(list(filter(lambda i: i == '0', ''.join(list(map(lambda i: "{0:b}".format(i), (filter(lambda i: i%2==0, numbers))))))))
   return 'odds win' if odd > even else 'evens win' if odd < even else 'tie'


def bits_battle(a):
   odds, evens = sum([bin(i).count('1') for i in a if i % 2]), sum([bin(i).count('0') - 1 for i in a if i % 2 == 0])
   return 'odds win' if odds > evens else 'evens win' if evens > odds else 'tie'


def bits_battle(numbers):
   x = sum(format(n, 'b').count('1') if n % 2 else -format(n, 'b').count('0') for n in numbers if n)
   return ( 'odds win' if x > 0 else 'evens win' if x < 0 else 'tie' )


def bits_battle(numbers):
   odds = [i for i in numbers if i % 2 ]
   evens = [i for i in numbers if i % 2 == 0]
   res = sum(bin(i).count('1') for i in odds) - sum(bin(i)[2:].count('0') for i in evens)
   return 'tie' if res == 0 else ['odds win', 'evens win'][res < 0]


def bits_battle(numbers):
   s = sum(map(lambda x: f'{x:b}'.count('1') if x%2 else -f'{x:b}'.count('0'), numbers))
   return {s>0:'odds win', s==0:'tie', s<0: 'evens win'}.get(True)


print(bits_battle([5, 3, 14]), 'odds win')
print(bits_battle([3, 8, 22, 15, 78]), 'evens win')
print(bits_battle([]), 'tie')
print(bits_battle([1, 13, 16]), 'tie')

File: JavaScript/7js 1/bitsBattle.js
/*
|--------------------------------------------------------------------------
| Bits Battle
|--------------------------------------------------------------------------
|
*/

function bitsBattle(numbers) {
   let odd = numbers.filter(i => i % 2 != 0).map(j => j.toString(2)).join('').split('').filter(i => i == '1').length;
   let even = numbers.filter(i => i % 2 == 0).map(j => j.toString(2)).join('').split('').filter(i => i == '0').length;
   return odd > even ? 'odds win' : odd < even ? 'evens win' : 'tie';
}


function bitsBattle(numbers) {
   let r = [0, 0]
   for (let n of numbers) r[n % 2] += n.toString(2).split(n % 2).length - 1
   return r[0] > r[1] ? "evens win" : r[1] > r[0] ? "odds win" : "tie"
}


const bitsBattle = arr => {
   let score = arr.reduce((a, x) => a + (x & 1 ? +1 : -1) * (x && x.toString(2).split(x & 1).length - 1), 0);
   return score > 0 ? "odds win" : score < 0 ? "evens win" : "tie";
}


const bitsBattle = numbers =>
   (val => val ? `${val < 0 ? `odds` : `evens`} win` : `tie`)
   (numbers.reduce((pre, val) => pre + (val & 1 ? -1 : 1) * (val.toString(2).split(val & 1).length - 1), 0));


function bitsBattle(numbers) {
   const odds = (numbers.filter(e => e % 2 != 0).map(e => e.toString(2)).join('').match(/1/g) || []).length;
   const evens = (numbers.filter(e => e % 2 == 0).map(e => e.toString(2)).join('').match(/0/g) || []).length;
   return evens > odds ? 'evens win' : evens == odds ? 'tie' : 'odds win'
}


console.log(bitsBattle([5, 3, 14]), 'odds win');
console.log(bitsBattle([3, 8, 22, 15, 78]), 'evens win');
console.log(bitsBattle([]), 'tie');
console.log(bitsBattle([1, 13, 16]), 'tie');